### PR TITLE
Add latency field in http_request.

### PR DESCRIPTION
**Pull request by qingling128**

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ Metrics/AbcSize:
 Metrics/ClassLength:
   Max: 1300
 Metrics/ModuleLength:
-  Max: 1400
```
Comment (qingling128) [inline .rubocop_todo.yml]: -___- :facepalm:

File: fluent-plugin-google-cloud.gemspec [hunk 2]
```diff
@@ -19,6 +19,7 @@ eos
   gem.test_files    = gem.files.grep(/^(test)/)
   gem.require_paths = ['lib']
 
+  gem.add_runtime_dependency 'chronic_duration', '~> 0.10.6'
```
Comment (qingling128) [inline fluent-plugin-google-cloud.gemspec]: This is a third party lib that parses strings to duration: https://github.com/henrypoydar/chronic_duration

Is it a good practice to use something like this? Maybe we can fork it if necessary. Or simply use a regex instead?

**Comment by qingling128** [comment]
@igorpeshansky Got a sec for this one?

File: fluent-plugin-google-cloud.gemspec (hunk 2 above)
Comment (igorpeshansky) [inline fluent-plugin-google-cloud.gemspec]: I would ask the Google Ruby libraries folks. This is a common issue with all of our APIs, and should have a common solution by now.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 3]
```diff
@@ -961,6 +964,27 @@ def set_http_request(record, entry)
...
+            # Split the integer and decimal parts in order to calculate seconds
+            # and nanos.
+            (integer_part, decimal_part) = duration_in_seconds.divmod 1
+            output.latency = Google::Protobuf::Duration.new(
+              seconds: integer_part,
+              nanos: (decimal_part * NANOS_IN_A_SECOND).round
+            )
+          else
+            # Logging API only takes a string with digits that ends with 's'.
+            output.latency = "#{duration_in_seconds.round(2)}s"
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: You can send a hash with `seconds` and `nanos` here as well, just like we do for the timestamp above.

Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Good to know. Changed.

File: fluent-plugin-google-cloud.gemspec (hunk 2 above)
Comment (qingling128) [inline fluent-plugin-google-cloud.gemspec]: Asked [Steven](https://github.com/sqrrrl) who is a major contributor of [google-api-ruby-client](https://github.com/google/google-api-ruby-client). It's OK to have dependencies on 3rd party libraries. It's a common thing in our open source projects and API clients. Note that the licenses do matter though. Affero General Public License ([AGPL](https://en.wikipedia.org/wiki/Affero_General_Public_License)) is not allowed.

**Comment by qingling128** [comment]
@igorpeshansky Got a sec for this one?

File: fluent-plugin-google-cloud.gemspec (hunk 2 above)
Comment (qingling128) [inline fluent-plugin-google-cloud.gemspec]: Per discussion offline, we decided to use simple regex instead of this third-party lib since it is not actively maintained. I just modified the implementation.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 4]
```diff
@@ -67,6 +66,8 @@ module Constants
         service: 'ml.googleapis.com',
         resource_type: 'ml_job'
       }
+
+      LATENCY_REGEX = /^[ \t]*(\d+)(\.\d+)?[ \t]*s[ \t]*$/
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Let's replace `[ \t]` with `\s`...

Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Instead of putting this into the `Constants` module, I'd suggest making this a class variable and moving it to `confgure`, alongside `@cloudfunctions_tag_regexp` and the like. Also, the naming convention in this file uses `regexp` instead of `regex`...

File: lib/fluent/plugin/out_google_cloud.rb [hunk 5]
```diff
@@ -967,12 +968,16 @@ def set_http_request(record, entry)
...
       unless latency.nil?
-        # Parse latency to duration. If failed, skip setting latency.
-        duration_in_seconds = ChronicDuration.parse(latency)
-        if duration_in_seconds
+        # Parse latency. If no valid format is detected, skip setting latency.
+        # Format: space/tab (optional) + integer + point & decimal (optional)
+        #       + space/tab (optional) + "s" + space/tab (optional)
+        # e.g.: "1.42 s"
+        parsed_groups = latency.scan(LATENCY_REGEX)
+        if !parsed_groups.empty? && parsed_groups[0].length == 2
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: You have exactly two captures in the string. If there's a match (`parsed_groups` is non-empty), when will `parsed_groups[0].length` be anything but 2?

File: lib/fluent/plugin/out_google_cloud.rb [hunk 6]
```diff
@@ -967,12 +968,16 @@ def set_http_request(record, entry)
 
       latency = input.delete('latency')
       unless latency.nil?
-        # Parse latency to duration. If failed, skip setting latency.
-        duration_in_seconds = ChronicDuration.parse(latency)
-        if duration_in_seconds
+        # Parse latency. If no valid format is detected, skip setting latency.
+        # Format: space/tab (optional) + integer + point & decimal (optional)
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: s/`space/tab`/`whitespace`/g

File: lib/fluent/plugin/out_google_cloud.rb [hunk 7]
```diff
@@ -967,12 +968,16 @@ def set_http_request(record, entry)
...
+        #       + space/tab (optional) + "s" + space/tab (optional)
+        # e.g.: "1.42 s"
+        parsed_groups = latency.scan(LATENCY_REGEX)
+        if !parsed_groups.empty? && parsed_groups[0].length == 2
           # Split the integer and decimal parts in order to calculate seconds
           # and nanos.
-          (latency_seconds, decimal_part) = duration_in_seconds.divmod 1
+          latency_seconds = parsed_groups[0][0].to_i
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: This might be a bit more readable if we change the regexp to `/^\s*(?<seconds>\d+)(?<decimal>\.\d+)?\s*s\s*$/`, and then use:
```
match = LATENCY_REGEX.match(latency)
if match
  latency_seconds = match['seconds'].to_i
  decimal_part = match['decimal'].to_f
  ...
end
```

File: test/plugin/base_test.rb [hunk 8]
```diff
@@ -1628,18 +1628,11 @@ def http_request_message_with_absent_referer
...
       '32s' => { 'seconds' => 32, 'nanos' => 0 },
-      '32 second' => { 'seconds' => 32, 'nanos' => 0 },
-      '32 seconds' => { 'seconds' => 32, 'nanos' => 0 },
-      '32 sec' => { 'seconds' => 32, 'nanos' => 0 },
-      '32 secs' => { 'seconds' => 32, 'nanos' => 0 },
       '0.32s' => { 'seconds' => 0, 'nanos' => 320_000_000 },
-      '123' => { 'seconds' => 123, 'nanos' => 0 },
-      '1.3442' => { 'seconds' => 1, 'nanos' => 344_200_000 }
+      ' 123 s ' => { 'seconds' => 123, 'nanos' => 0 },
+      '1.3442 s' => { 'seconds' => 1, 'nanos' => 344_200_000 }
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: You mentioned tabs in the latency format description, but aren't testing for those... With `\s`, we could also test for other whitespace...

Comment (qingling128) [inline test/plugin/base_test.rb]: Added.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 7 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Good idea. Changed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 6 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 4 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Changed the location and the name.

Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 5 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Yeah, it's kind of redundant. I just removed it. I was mainly worried what if someone mistakenly touches the regex and the length changes? It might crash the agent if the tests failed to catch it. Yet since we've changed the regex to have the groups named as `seconds` and `decimal`, I think the chance of such mistake should be quite low. I'll stop worrying about it.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 9]
```diff
@@ -244,6 +242,7 @@ def configure(conf)
         \[(?<timestamp>.{24})\]
         (?:\[(?<execution_id>[^\]]+)\])?
         [ ](?<text>.*)$/x
+      @http_latency_regexp = /^\s*(?<seconds>\d+)(?<decimal>\.\d+)?\s*s\s*$/
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Add a blank line before this, as it's a separate regexp group.

File: test/plugin/base_test.rb (hunk 8 above)
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: Can we make the tabs explicit, please? I.e., use `"\t\t123\t\ts\t\t"` instead of embedding tabs inline.
Also use `\r`, `\n`, `\v`, `\f`...

Comment (qingling128) [inline test/plugin/base_test.rb]: Added.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 9 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Added.

File: test/plugin/base_test.rb (hunk 8 above)
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: I would have just dumped them all in one test case, but this is fine too.

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 10]
```diff
@@ -152,6 +152,18 @@ def test_non_integer_timestamp
...
 
+  # The conversions from user input to output.
+  def latency_conversion
+    # TODO(lingshi): Use http://ruby-doc.org/core-2.4.0/Hash.html#method-i-transform_values
+    # instead when we upgrade to ruby 2.4.0.
+    # super.transform_values do |expected|
+    #   expected.reject { |_, value| value == 0 }
+    # end
+    super.map do |input, expected|
+      [input, expected.reject { |_, value| value == 0 }]
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Does the REST path require us to send 'nanos' => 0? If not, can we just get rid of those throughout?

Comment (qingling128) [inline test/plugin/test_out_google_cloud_grpc.rb]: Did some experiment. The REST path does not requrie `'nanos' => 0` to be set. Changed.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 11]
```diff
@@ -961,6 +965,30 @@ def set_http_request(record, entry)
...
+          # Split the integer and decimal parts in order to calculate seconds
+          # and nanos.
+          latency_seconds = match['seconds'].to_i
+          latency_nanos = (match['decimal'].to_f * NANOS_IN_A_SECOND).round
+          if @use_grpc
+            output.latency = Google::Protobuf::Duration.new
+          else
+            output.latency = {}
+          end
+          output.latency['seconds'] = latency_seconds \
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: This will happen automagically for protobufs, so I would just simplify the hash case:
```
            output.latency = {
              seconds: latency_seconds,
              nanos: latency_nanos
            }.delete_if { |k, v| v == 0 }
```

File: test/plugin/base_test.rb [hunk 12]
```diff
@@ -1368,6 +1368,39 @@ def test_http_request_from_record_with_referer_nil_or_absent
...
+        assert_equal HTTP_REQUEST_MESSAGE.merge('latency' => expected),
+                     entry['httpRequest'], entry
+        assert_nil get_fields(entry['jsonPayload'])['httpRequest'], entry
+      end
+    end
+  end
+
+  def test_http_request_with_null_latency
+    setup_gce_metadata_stubs
+    ['', nil, 'null'].each do |input|
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: How about also testing something that doesn't match the regex?

Comment (qingling128) [inline test/plugin/base_test.rb]: Added.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 11 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: test/plugin/base_test.rb [hunk 13]
```diff
@@ -1368,6 +1368,42 @@ def test_http_request_from_record_with_referer_nil_or_absent
...
+        assert_nil get_fields(entry['jsonPayload'])['httpRequest'], entry
+      end
+    end
+  end
+
+  # Skip setting latency when the field is null or with invalid format.
+  def test_http_request_skip_setting_latency
+    setup_gce_metadata_stubs
+    [
+      '', ' ', nil, 'null', '123', '1.23 seconds', '1min', 'abc&^!$*('
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: Let's also add `' 123  s     econds  '` or something like that.

File: test/plugin/base_test.rb [hunk 14]
```diff
@@ -1368,6 +1368,42 @@ def test_http_request_from_record_with_referer_nil_or_absent
...
+      end
+      verify_log_entries(1, COMPUTE_PARAMS, 'httpRequest') do |entry|
+        assert_equal HTTP_REQUEST_MESSAGE.merge('latency' => expected),
+                     entry['httpRequest'], entry
+        assert_nil get_fields(entry['jsonPayload'])['httpRequest'], entry
+      end
+    end
+  end
+
+  # Skip setting latency when the field is null or with invalid format.
```
Comment (igorpeshansky) [inline test/plugin/base_test.rb]: s/`with`/`has`/

File: test/plugin/base_test.rb (hunk 13 above)
Comment (qingling128) [inline test/plugin/base_test.rb]: Added.

File: test/plugin/base_test.rb (hunk 14 above)
Comment (qingling128) [inline test/plugin/base_test.rb]: Changed.
